### ISSUE TITLE
[Feature]: a layer can only own its KV-cache spec if it is DeepSeek-V4 or Mamba (hardcoded isinstance whitelist)

### ISSUE THREAD

**Issue by jt226ub** [vllm-project/tpu-inference#3538]
## Summary

`KVCacheManager.get_kv_cache_spec` (`tpu_inference/runner/kv_cache_manager.py:80`, `:563`) lets a
layer supply its own spec through
`attn_module.get_kv_cache_spec(vllm_config)` — but only if the layer passes one of two
predicates: `isinstance(attn_module, MambaBase)`, or `is_cache_for_ds_v4(attn_module)`, which is
an `isinstance` whitelist of four concrete DeepSeek-V4 classes. Any other architecture with a
non-standard cache — an indexer, a compressor, a ring/sliding side cache — cannot register a
spec without editing `kv_cache_manager.py`.

The two branches are **byte-identical in body**; only the predicate differs. One is a protocol,
the other is a list of four class names.

## Evidence

`tpu_inference/runner/kv_cache_manager.py:69-73`:

```python
def is_cache_for_ds_v4(attn_module: AttentionLayerBase) -> bool:
    return isinstance(attn_module, DeepseekV4IndexerCache) or isinstance(
        attn_module, DeepseekV4SWACache) or isinstance(
            attn_module, DeepseekV4Attention) or isinstance(
                attn_module, CompressorStateCache)
```

`:727-739` — the only call site, and the duplicated body:

```python
            for layer_name, attn_module in layers.items():
                if isinstance(attn_module, MambaBase):
                    spec = attn_module.get_kv_cache_spec(self.runner.vllm_config)
                    if spec is not None:
                        kv_cache_spec[layer_name] = spec
                    continue

                if is_cache_for_ds_v4(attn_module):
                    spec = attn_module.get_kv_cache_spec(self.runner.vllm_config)
                    if spec is not None:
                        kv_cache_spec[layer_name] = spec
                    continue
```

Everything below that point assumes a standard attention geometry.

## Who this blocks

**Independently reported in #3081**: a DeepSeek-V4 dp-attention configuration where the reporter
hotpatched the compressor-state layer to carry its own block size, and hit the assert at
`kv_cache_manager.py:899-903` — reached from `initialize_kv_cache` (`:825`), not from
`get_kv_cache_spec` — which requires every non-MTP `kv_cache_tensor` to carry the **same number
of shared layers**:

```python
                        num_shared_layers = len(non_mtp_tensors[0].layers)
                        for kv_cache_tensor in non_mtp_tensors:
                            assert len(
                                kv_cache_tensor.layers
                            ) == num_shared_layers, ...
```

A layer that owns a differently-shaped spec lands in a tensor with a different share count and
trips it. Same root cause: spec ownership is not a
first-class capability, so any deviation is a patch rather than an implementation.

**Downstream port** (ours): serving `qwen4_exp` (Qwen3.8-Flash-Next, hybrid gated attention with
a per-request raw-key ring and a compressed-key side cache) required three edits to
`kv_cache_manager.py` to let 12 attention layers own two extra spec types. Every one of those
edits is a line that upstream would have to carry per architecture. With the whitelist, the
work is O(architectures); with a protocol, it is O(1).

## Suggested fix

Replace both predicates with one capability check. The bodies are already identical — but the
predicate needs one exclusion, and it is not obvious:

```python
            for layer_name, attn_module in layers.items():
                if is_kv_cache_spec_owner(attn_module):
                    spec = attn_module.get_kv_cache_spec(self.runner.vllm_config)
                    if spec is not None:
                        kv_cache_spec[layer_name] = spec
                    continue
```

where the predicate is *"declares its own `get_kv_cache_spec` **and** is not one of the layer
types the standard path already prices"*:

```python
def is_kv_cache_spec_owner(attn_module) -> bool:
    if isinstance(attn_module, (Attention, MLAAttention)):
        return False          # priced by the standard path below, with padding and sharding
    return hasattr(attn_module, "get_kv_cache_spec")
```

**A bare `hasattr` (or a bare `Protocol` `isinstance`) is unsafe**, which we found by building it:
`get_kv_cache_spec` is an `@abstractmethod` on `AttentionLayerBase`
(`vllm/model_executor/layers/attention/../attention_layer_base.py:34-44`), so *every*
attention-like layer satisfies the protocol structurally — including ordinary
`Attention`. Without the exclusion, every standard attention layer is diverted away from the
padding and sharding the runner applies at `kv_cache_manager.py:755-776`. That is a silent
behaviour change for every currently-supported model, i.e. the opposite of reviewable.

With the exclusion it is behaviour-preserving, and verifiably so: the four whitelisted
DeepSeek-V4 classes subclass `AttentionLayerBase` **directly**, never `Attention` —
`vllm/models/deepseek_v4/attention.py:131` (`DeepseekV4Attention(nn.Module, AttentionLayerBase, ABC)`),
`:862` (`DeepseekV4IndexerCache(torch.nn.Module, AttentionLayerBase)`),
`vllm/models/deepseek_v4/compressor.py:133` (`CompressorStateCache`), and
`vllm/v1/attention/backends/mla/sparse_swa.py:64` (`DeepseekV4SWACache`) — while
`Attention` and `MLAAttention` are separate `AttentionLayerBase` subclasses
(`layers/attention/attention.py:223`, `layers/attention/mla_attention.py:402`). So the exclusion
changes nothing for the layers the whitelist currently admits, and the diff still deletes a
whitelist and a duplicated block.

We have this implemented and tested offline against upstream source at `623e68a14`: a
third-party layer type with a differently-shaped spec (block 64, 1 head, head_size 128, against
standard layers at 2×64) gets its spec honoured intact, and the whole
`kv_cache_spec` dict for `{mamba, indexer, compressor, attn.0, attn.1}` is byte-identical before
and after. Happy to send it as a PR.

Two follow-ups that the protocol does not by itself solve, listed so the scope is honest:

1. **Page-size grouping still assumes one geometry** — the `:899-903` assert #3081 hits, and the
   uniform-geometry assert in `update_mamba_page_size_padded` (`:147-296`). Filed separately /
   discussed in #3483.
2. **`merge` must reject foreign specs cleanly.** `_get_kv_cache_groups_uniform_page_size`
   probes buckets with `type(specs[0]).merge([*specs, layer_spec])` and relies on the exception
   type; a custom spec whose `merge` raises the wrong exception fails engine init rather than
   forming a new group. Worth documenting as part of the spec-owner contract — it cost us a
   whole device iteration to discover.

## What we are asking for

Confirmation that a generic protocol is the direction you want, before anyone writes it. If yes
we will send the patch (protocol + whitelist deletion + a test that a third-party layer type
gets its spec honoured without touching `kv_cache_manager.py`). If you would rather keep an
explicit registry — e.g. layers register a spec factory at import — that works for us too; the
requirement is only that it not be an `isinstance` list in the runner.

**Comment by jt226ub** [comment]
Cross-references: the grouping half of this is #3541 (one `MambaSpec` forces per-layer caches, then asserts a uniform shared-layer count), and the page-pricing half is discussed in #3483. #3081 is the third-party report that dies on the #3541 assert after doing what this issue asks for.
